cns/restserver/ipam.go

```diff
@@ -478,101 +577,129 @@ func (service *HTTPRestService) MarkExistingIPsAsPendingRelease(pendingIPIDs []s
+	IPMap := make(map[string]string)
+	for _, desiredIP := range desiredIPAddresses {
+		IPMap[desiredIP] = desiredIP
+	}
+	numOfIPs := 0
 
-	for _, ipConfig := range service.PodIPConfigState {
-		if ipConfig.IPAddress == desiredIPAddress {
+forLoop:
+	for _, ipConfig := range service.PodIPConfigState { //nolint:gocritic // ignore copy
+		if _, found := IPMap[ipConfig.IPAddress]; found {
 			switch ipConfig.GetState() { //nolint:exhaustive // ignoring PendingRelease case intentionally
 			case types.Assigned:
 				// This IP has already been assigned, if it is assigned to same pod, then return the same
 				// IPconfiguration
 				if ipConfig.PodInfo.Key() == podInfo.Key() {
-					logger.Printf("[AssignDesiredIPConfig]: IP Config [%+v] is already assigned to this Pod [%+v]", ipConfig, podInfo)
+					logger.Printf("[AssignDesiredIPConfigs]: IP Config [%+v] is already assigned to this Pod [%+v]", ipConfig, podInfo)
 				} else {
-					return podIpInfo, errors.Errorf("[AssignDesiredIPConfig] Desired IP is already assigned %+v, requested for pod %+v", ipConfig, podInfo)
+					logger.Errorf("[AssignDesiredIPConfigs] Desired IP is already assigned %+v, requested for pod %+v", ipConfig, podInfo)
+					break forLoop
 				}
 			case types.Available, types.PendingProgramming:
 				// This race can happen during restart, where CNS state is lost and thus we have lost the NC programmed version
 				// As part of reconcile, we mark IPs as Assigned which are already assigned to Pods (listed from APIServer)
 				if err := service.assignIPConfig(ipConfig, podInfo); err != nil {
-					return podIpInfo, err
+					logger.Errorf(err.Error())
+					break forLoop
 				}
 			default:
-				return podIpInfo, errors.Errorf("[AllocateDesiredIPConfig] Desired IP is not available %+v", ipConfig)
+				logger.Errorf("[AllocateDesiredIPConfig] Desired IP is not available %+v", ipConfig)
+				break forLoop
+			}
+			err := service.populateIPConfigInfoUntransacted(ipConfig, &podIPInfo[numOfIPs])
+			numOfIPs++
+			if numOfIPs == len(desiredIPAddresses) {
+				return podIPInfo, err
 			}
-			err := service.populateIPConfigInfoUntransacted(ipConfig, &podIpInfo)
-			return podIpInfo, err
 		}
 	}
-	return podIpInfo, fmt.Errorf("Requested IP not found in pool")
+	//nolint:goerr113 // return error
+	return podIPInfo, fmt.Errorf("Not all requested IPs were found/available in the pool")
 }
 
-func (service *HTTPRestService) AssignAnyAvailableIPConfig(podInfo cns.PodInfo) (cns.PodIpInfo, error) {
+// Assigns an IP from each NC on the NNC
+func (service *HTTPRestService) AssignAvailableIPConfigs(podInfo cns.PodInfo) ([]cns.PodIpInfo, error) {
 	service.Lock()
 	defer service.Unlock()
+	// Creates a slice of PoPodIpInfo with the size of number of NCs
+	podIPInfo := make([]cns.PodIpInfo, len(service.state.ContainerStatus))
+	// This map is used to store whether or not we have found an available IP from an NC when looping through the pool
+	ncMap := make(map[string]any)
```

Comment:
when using a map as a set, the value type should be empty struct 
```suggestion
	ncMap := make(map[string]struct{})
```

Reply:
Changed to the struct I am putting into the map instead of any